src/main/cpp/src/NativeParquetJni.cpp

```diff
@@ -104,217 +116,285 @@ class column_pruner {
+          throw std::runtime_error("found non repeating map child");
+        }
+
+        int repeated_field_num_children = get_num_children(repeated_field_schema_item);
 
-      return column_pruning_maps{std::move(final_schema_map),
-          std::move(final_num_children_map),
-          std::move(final_chunk_map)};
+        if (repeated_field_num_children != 1 && repeated_field_num_children != 2) {
+          throw std::runtime_error("found map with wrong number of children");
+        }
+
+        schema_map.push_back(current_input_schema_index);
+        schema_num_children.push_back(repeated_field_num_children);
+        ++current_input_schema_index;
+
+        // Process the key...
+        key_found.filter_schema(schema, ignore_case, current_input_schema_index, next_input_chunk_index, chunk_map, schema_map, schema_num_children);
+        if (repeated_field_num_children == 2) {
+          // Process the value...
+          value_found.filter_schema(schema, ignore_case, current_input_schema_index, next_input_chunk_index, chunk_map, schema_map, schema_num_children);
+        }
     }
 
-private:
+    /**
+     * Recursive method to parse and update the maps to filter out columns in the schema and chunks.
+     * Each column_pruner is responsible to parse out from schema what it holds and skip anything
+     * that does not match. chunk_map, schema_map, and schema_num_children are the final outputs.
+     * current_input_schema_index and next_input_chunk_index are also outputs but are state that is
+     * passed to each child and returned when it comsumes comething.
+     */
+    void filter_schema(std::vector<parquet::format::SchemaElement> & schema, bool const ignore_case,
+            std::size_t & current_input_schema_index, std::size_t & next_input_chunk_index,
+            std::vector<int> & chunk_map, std::vector<int> & schema_map, std::vector<int> & schema_num_children) const {
+      switch(tag) {
+        case Tag::STRUCT:
+          filter_schema_struct(schema, ignore_case, current_input_schema_index, next_input_chunk_index, chunk_map, schema_map, schema_num_children);
+          break;
+        case Tag::VALUE:
+          filter_schema_value(schema, current_input_schema_index, next_input_chunk_index, chunk_map, schema_map, schema_num_children);
+          break;
+        case Tag::LIST:
+          filter_schema_list(schema, ignore_case, current_input_schema_index, next_input_chunk_index, chunk_map, schema_map, schema_num_children);
+          break;
+        case Tag::MAP:
+          filter_schema_map(schema, ignore_case, current_input_schema_index, next_input_chunk_index, chunk_map, schema_map, schema_num_children);
+          break;
+        default:
+          throw std::runtime_error(std::string("INTERNAL ERROR UNEXPECTED TAG FOUND ") + std::to_string(static_cast<int>(tag)));
+      }
+    }
 
+    /**
+     * Do a depth first traversal to build up this.
```

Comment:
```suggestion
     * Do a depth first traversal to build up schema elements.
```
Is this building schema elements? Can we replace "this" with something more descriptive?